examples/virtual_methods/base_button/mod.rs

```diff
@@ -0,0 +1,110 @@
+/// Public part of the BaseButton
+mod imp;
+
+pub use self::imp::PinnedFuture;
+use gtk::glib::{
+    self,
+    subclass::prelude::*,
+    translate::{from_glib_borrow, ToGlibPtr},
+};
+use gtk::{prelude::*, subclass::prelude::*};
+
+glib::wrapper! {
+    pub struct BaseButton(ObjectSubclass<imp::BaseButton>)
+        @extends gtk::Widget, gtk::Button;
+}
+
+/// Public trait that implements our functions for everything that derives from BaseButton
+pub trait BaseButtonExt {
+    fn sync_method(&self, extra_text: Box<Option<String>>);
+    fn async_method(&self) -> PinnedFuture;
+}
+
+/// We call into imp::BaseButton_$method_name for each function. These will retrieve the
+/// correct class (the base class for the BaseButton or the derived class for DerivedButton)
+/// and call the correct implementation of the function.
+impl<O: IsA<BaseButton>> BaseButtonExt for O {
+    fn sync_method(&self, extra_text: Box<Option<String>>) {
+        unsafe {
+            imp::base_button_sync_method(self.as_ref().to_glib_none().0, Box::into_raw(extra_text))
+        }
+    }
+
+    fn async_method(&self) -> PinnedFuture {
+        unsafe { imp::base_button_async_method(self.as_ref().to_glib_none().0) }
+    }
+}
+
+/// The BaseButtonImpl that each derived private struct has to implement. See derived_button/imp.rs for how
+/// to override functions.
+pub trait BaseButtonImpl: ButtonImpl + ObjectImpl + 'static {
+    fn sync_method(&self, obj: &BaseButton, extra_text: Box<Option<String>>) {
+        self.parent_sync_method(obj, extra_text)
+    }
+
+    fn async_method(&self, obj: &BaseButton) -> PinnedFuture {
+        self.parent_async_method(obj)
+    }
+
+    fn parent_sync_method(&self, obj: &BaseButton, extra_text: Box<Option<String>>) {
```

Comment:
The `parent_` methods should be moved into an extension trait that is only implemented for `BaseButtonImpl`s like in the bindings. Subclasses must not really override these.